JSON encoding of optional map field causes compiler error

### Describe the bug

If a struct that is being json encoded has a field which has a ?map[string]... type, a compiler bug occurs.

### Expected Behavior

Correctly encoding the struct, and omitting the field from encoding if the optional is `none`, just like when other optionals are encoded.

### Current Behavior
```
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2362:55: warning: incompatible pointer types passing '_option_string *' (aka 'struct _option_string *') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&(string[]){ json__decode_string(root) }, &res, sizeof(string));
                                                             ^~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1790:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2365:24: warning: incompatible pointer types passing 'byte (*)[56]' (aka 'unsigned char (*)[56]') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&res.data, &ret.data, sizeof(string));
                              ^~~~~~~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1790:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2416:70: warning: incompatible pointer types passing '_option_Array_string *' (aka 'struct _option_Array_string *') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&(Array_string[]) { __new_array(0, 0, sizeof(string)) }, &res, sizeof(Array_string));
                                                                            ^~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1790:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2428:24: warning: incompatible pointer types passing 'byte (*)[72]' (aka 'unsigned char (*)[72]') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&res.data, &ret.data, sizeof(Array_string));
                              ^~~~~~~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1790:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2481:6: error: assigning to '_option_Map_string_string' (aka 'struct _option_Map_string_string') from incompatible type 'map' (aka 'struct map')
        res = new_map(sizeof(string), sizeof(string), &map_hash_string, &map_eq_string, &map_clone_string, &map_free_string);
            ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2487:11: warning: incompatible pointer types passing '_option_Map_string_string *' (aka 'struct _option_Map_string_string *') to parameter of type 'map *' (aka 'struct map *') [-Wincompatible-pointer-types]
                map_set(&res, &key, &val);
                        ^~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1755:19: note: passing argument to parameter 'm' here
void map_set(map* m, voidptr key, voidptr value);
                  ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2492:24: warning: incompatible pointer types passing 'byte (*)[160]' (aka 'unsigned char (*)[160]') to parameter of type '_option *' (aka 'struct _option *') [-Wincompatible-pointer-types]
        _option_ok(&res.data, &ret.data, sizeof(Map_string_string));
                              ^~~~~~~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1790:40: note: passing argument to parameter 'option' here
void _option_ok(voidptr data, _option* option, int size);
                                       ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2501:30: warning: incompatible pointer types passing '_option_Map_string_string *' (aka 'struct _option_Map_string_string *') to parameter of type 'map *' (aka 'struct map *') [-Wincompatible-pointer-types]
        Array_string _t9 = map_keys(&val);
                                    ^~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1766:21: note: passing argument to parameter 'm' here
array map_keys(map* m);
                    ^
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:2504:86: warning: incompatible pointer types passing '_option_Map_string_string *' (aka 'struct _option_Map_string_string *') to parameter of type 'map *' (aka 'struct map *') [-Wincompatible-pointer-types]
                cJSON_AddItemToObject(o, (char*) key.str, json__encode_string ( *(string*) map_get(&val, &key, &(string[]) { (string){.str=(byteptr)"", .is_lit=1} } ) ) );
                                                                                                   ^~~~
/tmp/v_501/optionalmap.10614931672008338789.tmp.c:1761:22: note: passing argument to parameter 'm' here
voidptr map_get(map* m, voidptr key, voidptr zero);
                     ^
8 warnings and 1 error generated.
builder error: 
==================
C error. This should never happen.
```
### Reproduction Steps

```
module main

import json

struct Test {
	optional_string ?string //works
	optional_array ?[]string //works
	optional_struct_array ?[]string //works
	optional_map ?map[string]string //causes error
}

fn main() {
	test := Test {}
	encoded := json.encode(test)
}
```

### Possible Solution

perhaps implement encoding the way it is implemented for optional arrays?

### Additional Information/Context

_No response_

### V version

0.3.3

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 f1e9a8f.adcd16b
OS: macos, macOS, 13.2.1, 22D68
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz

--- a/vlib/json/json_encode_map_test.v
+++ b/vlib/json/json_encode_map_test.v
@@ -1,0 +1,22 @@
+import json
+
+struct Test {
+	optional_string       ?string
+	optional_array        ?[]string
+	optional_struct_array ?[]string
+	optional_map          ?map[string]string
+}
+
+fn test_main() {
+	test := Test{}
+	encoded := json.encode(test)
+	assert dump(encoded) == '{}'
+
+	test2 := Test{
+		optional_map: {
+			'foo': 'bar'
+		}
+	}
+	encoded2 := json.encode(test2)
+	assert dump(encoded2) == '{"optional_map":{"foo":"bar"}}'
+}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -135,8 +135,8 @@ ${enc_fn_dec} {
 			m := sym.info as ast.Map
 			g.gen_json_for_type(m.key_type)
 			g.gen_json_for_type(m.value_type)
-			dec.writeln(g.decode_map(m.key_type, m.value_type, ret_styp))
-			enc.writeln(g.encode_map(m.key_type, m.value_type))
+			dec.writeln(g.decode_map(utyp, m.key_type, m.value_type, ret_styp))
+			enc.writeln(g.encode_map(utyp, m.key_type, m.value_type))
 		} else if sym.kind == .alias {
 			a := sym.info as ast.Alias
 			parent_typ := a.parent_type
@@ -156,8 +156,8 @@ ${enc_fn_dec} {
 				m := psym.info as ast.Map
 				g.gen_json_for_type(m.key_type)
 				g.gen_json_for_type(m.value_type)
-				dec.writeln(g.decode_map(m.key_type, m.value_type, ret_styp))
-				enc.writeln(g.encode_map(m.key_type, m.value_type))
+				dec.writeln(g.decode_map(utyp, m.key_type, m.value_type, ret_styp))
+				enc.writeln(g.encode_map(utyp, m.key_type, m.value_type))
 			} else {
 				verror('json: ${sym.name} is not struct')
 			}
@@ -185,7 +185,7 @@ ${enc_fn_dec} {
 		dec.writeln('\t${result_name}_${ret_styp} ret;')
 		dec.writeln('\t_result_ok(&res, (${result_name}*)&ret, sizeof(res));')
 		if utyp.has_flag(.option) {
-			dec.writeln('\t_option_ok(&res.data, &ret.data, sizeof(${g.base_type(utyp)}));')
+			dec.writeln('\t_option_ok(&res.data, (${option_name}*)&ret.data, sizeof(${g.base_type(utyp)}));')
 		}
 		dec.writeln('\treturn ret;\n}')
 		enc.writeln('\treturn o;\n}')
@@ -242,7 +242,7 @@ fn (mut g Gen) gen_enum_enc_dec(utyp ast.Type, sym ast.TypeSymbol, mut enc strin
 		if is_option {
 			base_typ := g.typ(utyp.clear_flag(.option))
 			enc.writeln('\to = ${js_enc_name('u64')}(*val.data);')
-			dec.writeln('\t_option_ok(&(${base_typ}[]){ ${js_dec_name('u64')}(root) }, &res, sizeof(${base_typ}));')
+			dec.writeln('\t_option_ok(&(${base_typ}[]){ ${js_dec_name('u64')}(root) }, (${option_name}*)&res, sizeof(${base_typ}));')
 		} else {
 			dec.writeln('\tres = ${js_dec_name('u64')}(root);')
 			enc.writeln('\to = ${js_enc_name('u64')}(val);')
@@ -271,7 +271,7 @@ fn (mut g Gen) gen_option_enc_dec(typ ast.Type, mut enc strings.Builder, mut dec
 	enc.writeln('\to = ${encode_name}(*(${type_str}*)val.data);')
 
 	dec_name := js_dec_name(type_str)
-	dec.writeln('\t_option_ok(&(${type_str}[]){ ${dec_name}(root) }, &res, sizeof(${type_str}));')
+	dec.writeln('\t_option_ok(&(${type_str}[]){ ${dec_name}(root) }, (${option_name}*)&res, sizeof(${type_str}));')
 }
 
 [inline]
@@ -384,7 +384,7 @@ fn (mut g Gen) gen_sumtype_enc_dec(utyp ast.Type, sym ast.TypeSymbol, mut enc st
 				dec.writeln('\t\t${variant_typ} value = *(${variant_typ}*)(${tmp}.data);')
 			}
 			if is_option {
-				dec.writeln('\t\t\t_option_ok(&(${sym.cname}[]){ ${variant_typ}_to_sumtype_${sym.cname}(&value) }, &res, sizeof(${sym.cname}));')
+				dec.writeln('\t\t\t_option_ok(&(${sym.cname}[]){ ${variant_typ}_to_sumtype_${sym.cname}(&value) }, (${option_name}*)&res, sizeof(${sym.cname}));')
 			} else {
 				dec.writeln('\t\tres = ${variant_typ}_to_sumtype_${ret_styp}(&value);')
 			}
@@ -797,7 +797,7 @@ fn (mut g Gen) decode_array(utyp ast.Type, value_type ast.Type, fixed_array_size
 			fixed_array_idx_increment += 'fixed_array_idx++;'
 		} else {
 			array_element_assign += 'array_push${noscan}((array*)&res.data, &val);'
-			res_str += '_option_ok(&(${g.base_type(utyp)}[]) { __new_array${noscan}(0, 0, sizeof(${styp})) }, &res, sizeof(${g.base_type(utyp)}));'
+			res_str += '_option_ok(&(${g.base_type(utyp)}[]) { __new_array${noscan}(0, 0, sizeof(${styp})) }, (${option_name}*)&res, sizeof(${g.base_type(utyp)}));'
 			array_free_str += 'array_free(&res.data);'
 		}
 	} else {
@@ -873,7 +873,7 @@ fn (mut g Gen) encode_array(utyp ast.Type, value_type ast.Type, fixed_array_size
 '
 }
 
-fn (mut g Gen) decode_map(key_type ast.Type, value_type ast.Type, ustyp string) string {
+fn (mut g Gen) decode_map(utyp ast.Type, key_type ast.Type, value_type ast.Type, ustyp string) string {
 	styp := g.typ(key_type)
 	mut styp_v := g.typ(value_type)
 	ret_styp := styp_v.replace('*', '_ptr')
@@ -893,22 +893,39 @@ fn (mut g Gen) decode_map(key_type ast.Type, value_type ast.Type, ustyp string) 
 		${styp_v} val = *(${styp_v}*)val2.data;
 '
 	}
-	return '
-	if(!cJSON_IsObject(root) && !cJSON_IsNull(root)) {
-		return (${result_name}_${ustyp}){ .is_error = true, .err = _v_error(string__plus(_SLIT("Json element is not an object: "), tos2((byteptr)cJSON_PrintUnformatted(root)))), .data = {0}};
-	}
-	res = new_map(sizeof(${styp}), sizeof(${styp_v}), ${hash_fn}, ${key_eq_fn}, ${clone_fn}, ${free_fn});
-	cJSON *jsval = NULL;
-	cJSON_ArrayForEach(jsval, root)
-	{
-		${s}
-		string key = tos2((byteptr)jsval->string);
-		map_set(&res, &key, &val);
-	}
+
+	if utyp.has_flag(.option) {
+		return '
+		if(!cJSON_IsObject(root) && !cJSON_IsNull(root)) {
+			return (${result_name}_${ustyp}){ .is_error = true, .err = _v_error(string__plus(_SLIT("Json element is not an object: "), tos2((byteptr)cJSON_PrintUnformatted(root)))), .data = {0}};
+		}
+		_option_ok(&(${g.base_type(utyp)}[]) { new_map(sizeof(${styp}), sizeof(${styp_v}), ${hash_fn}, ${key_eq_fn}, ${clone_fn}, ${free_fn}) }, (${option_name}*)&res, sizeof(${g.base_type(utyp)}));
+		cJSON *jsval = NULL;
+		cJSON_ArrayForEach(jsval, root)
+		{
+			${s}
+			string key = tos2((byteptr)jsval->string);
+			map_set((map*)res.data, &key, &val);
+		}
 '
+	} else {
+		return '
+		if(!cJSON_IsObject(root) && !cJSON_IsNull(root)) {
+			return (${result_name}_${ustyp}){ .is_error = true, .err = _v_error(string__plus(_SLIT("Json element is not an object: "), tos2((byteptr)cJSON_PrintUnformatted(root)))), .data = {0}};
+		}
+		res = new_map(sizeof(${styp}), sizeof(${styp_v}), ${hash_fn}, ${key_eq_fn}, ${clone_fn}, ${free_fn});
+		cJSON *jsval = NULL;
+		cJSON_ArrayForEach(jsval, root)
+		{
+			${s}
+			string key = tos2((byteptr)jsval->string);
+			map_set(&res, &key, &val);
+		}
+'
+	}
 }
 
-fn (mut g Gen) encode_map(key_type ast.Type, value_type ast.Type) string {
+fn (mut g Gen) encode_map(utyp ast.Type, key_type ast.Type, value_type ast.Type) string {
 	styp := g.typ(key_type)
 	styp_v := g.typ(value_type)
 	fn_name_v := js_enc_name(styp_v)
@@ -921,13 +938,25 @@ fn (mut g Gen) encode_map(key_type ast.Type, value_type ast.Type) string {
 		// key += '${styp}_str((($styp*)${keys_tmp}.data)[i]);'
 		verror('json: encode only maps with string keys')
 	}
-	return '
-	o = cJSON_CreateObject();
-	Array_${styp} ${keys_tmp} = map_keys(&val);
-	for (int i = 0; i < ${keys_tmp}.len; ++i) {
-		${key}
-		cJSON_AddItemToObject(o, (char*) key.str, ${fn_name_v} ( *(${styp_v}*) map_get(&val, &key, &(${styp_v}[]) { ${zero} } ) ) );
-	}
-	array_free(&${keys_tmp});
+	if utyp.has_flag(.option) {
+		return '
+		o = cJSON_CreateObject();
+		Array_${styp} ${keys_tmp} = map_keys((map*)val.data);
+		for (int i = 0; i < ${keys_tmp}.len; ++i) {
+			${key}
+			cJSON_AddItemToObject(o, (char*) key.str, ${fn_name_v} ( *(${styp_v}*) map_get((map*)val.data, &key, &(${styp_v}[]) { ${zero} } ) ) );
+		}
+		array_free(&${keys_tmp});
 '
+	} else {
+		return '
+		o = cJSON_CreateObject();
+		Array_${styp} ${keys_tmp} = map_keys(&val);
+		for (int i = 0; i < ${keys_tmp}.len; ++i) {
+			${key}
+			cJSON_AddItemToObject(o, (char*) key.str, ${fn_name_v} ( *(${styp_v}*) map_get(&val, &key, &(${styp_v}[]) { ${zero} } ) ) );
+		}
+		array_free(&${keys_tmp});
+'
+	}
 }

